### PR TITLE
refactor(types): enable noImplicitAny instance.ts

### PR DESCRIPTION
**Pull request by laljikanjareeya**
Fixes #<issue_number_goes_here> (it's a good idea to open an issue first for discussion)

- [ ] Tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -30,6 +31,7 @@
     "spanner"
   ],
   "scripts": {
+    "alltests": "npm run test && npm run system-test",
```
Comment (AVaksman) [inline package.json]: I do not think we need this for the PR

File: package.json [hunk 2]
```diff
@@ -41,14 +43,11 @@
     "ycsb": "node ./benchmark/ycsb.js  run -P ./benchmark/workloada -p table=usertable -p cloudspanner.instance=ycsb-instance -p operationcount=100 -p cloudspanner.database=ycsb",
     "fix": "gts fix && eslint --fix '**/*.js'",
     "clean": "gts clean",
-    "compile": "tsc -p . && cp -r src/v1 build/src && cp -r protos build && cp test/*.js build/test",
+    "compile": "tsc -p . && cp -r src/v1 build/src && cp -r protos build && cp -r proto build && cp test/*.js build/test",
```
Comment (AVaksman) [inline package.json]: Do you need the repeating `-r proto build && cp `?

File: src/transaction.ts [hunk 3]
```diff
@@ -83,7 +84,7 @@ export interface ReadRequest extends RequestOptions {
   partitionToken?: Uint8Array | string;
 }
 
-export interface BatchUpdateError extends ServiceError {
+export interface BatchUpdateError extends RowCountsServiceError {
```
Comment (AVaksman) [inline src/transaction.ts]: Why change from `ServiceError` to `RowCountsServiceError`?
`RowCountsServiceError` already has `rowCounts` property.

File: src/common.ts [hunk 4]
```diff
@@ -8,11 +10,18 @@ export interface TransactionOptions {
...
-  seconds: number;
+  nanos: number; seconds: number;
+};
+
+export interface RowCountsServiceError {
+  //tslint:disable-next-line no-any
+  rowCounts?: number[] | ((rowCounts: any, arg1: never[]) => any);
+  code?: status;
+  metadata?: Metadata;
+  details?: string;
```
Comment (AVaksman) [inline src/common.ts]: if you extend from `ServiceError`, then you do need to re-introduce `code`, `metadata` and `details` properties. They are present in `ServiceError` interface.

File: src/common.ts [hunk 5]
```diff
@@ -1,3 +1,5 @@
+import { status, Metadata, ServiceError } from 'grpc';
```
Comment (AVaksman) [inline src/common.ts]: Use `ServiceError` to extend from below. And you won't need to import `status` and `metadata` here I believe

File: src/instance.ts [hunk 6]
```diff
@@ -41,7 +108,8 @@ import {Database} from './database';
  * const instance = spanner.instance('my-instance');
  */
 class Instance extends common.ServiceObject {
-  constructor(spanner, name) {
+  // "spanner: Spanner" causes flood of errors in test/instance
```
Comment (AVaksman) [inline src/instance.ts]: remove this line please

File: src/instance.ts [hunk 7]
```diff
@@ -249,7 +319,8 @@ class Instance extends common.ServiceObject {
    * const instance = spanner.instance('my-instance');
    * const database = instance.database('my-database');
    */
-  database(name, poolOptions?) {
+  // Return annotation of "Database" causes flood of errors in test/instance & system-test/spanner
+  database(name: string, poolOptions?: SessionPoolOptions | SessionPoolInterface) {
```
Comment (AVaksman) [inline src/instance.ts]: Remove the line above please.
This method does return a `Database`.

File: src/instance.ts [hunk 8]
```diff
@@ -16,13 +16,80 @@
...
+export interface CreateDatabaseCallback {
+  (err: Error | null, database?: Database | null, operation?: GaxOperation | null, apiResponse?: LongrunningIOperation): void;
+}
+export interface DeleteInstanceCallback {
+  (err: Error | null, apiResponse: dbInstanceAdminClient.protobuf.IEmpty): void;
+}
+export interface ExistsInstanceCallback {
+  (err: Error | null, exists: boolean | null): void;
+}
+export interface GetDatabasesCallback {
```
Comment (AVaksman) [inline src/instance.ts:99]: This interface does not conform to `GetDatabasesResponse` interface. Both `Response` and `Callback` should return the same items

File: src/transaction-runner.ts [hunk 9]
```diff
@@ -77,7 +78,7 @@ interface ErrorCallback {
  *
  * @param {Error} [err] The last known retryable Error.
  */
-export class DeadlineError extends Error implements ServiceError {
+export class DeadlineError extends Error implements RowCountsServiceError {
```
Comment (AVaksman) [inline src/transaction-runner.ts]: Why switch to `RowCountsServiceError`?

**Review by AVaksman — CHANGES_REQUESTED** [review]

**Comment by AVaksman** [comment]
Please make sure to `npm install` and `npm run fix`
This will remove a lot of formatting changes that are not related to this PR

File: src/instance.ts [hunk 10]
```diff
@@ -576,7 +660,8 @@ class Instance extends common.ServiceObject {
    *   const apiResponse = data[1];
    * });
    */
-  getMetadata(callback) {
+  //tslint:disable-next-line no-any
+  getMetadata(callback: GetInstanceMetadataCallback): any {
```
Comment (AVaksman) [inline src/instance.ts]: I think you can use `Metadata` instead of `any` here

**Review by AVaksman — CHANGES_REQUESTED** [review]

File: src/common.ts [hunk 11]
```diff
@@ -8,11 +10,18 @@ export interface TransactionOptions {
...
 }
 export interface GetTimestamp {
-  nanos: number;
-  seconds: number;
+  nanos: number; seconds: number;
+};
+
+export interface RowCountsServiceError {
+  //tslint:disable-next-line no-any
+  rowCounts?: number[] | ((rowCounts: any, arg1: never[]) => any);
```
Comment (AVaksman) [inline src/common.ts]: Where is `rowCounts` used?

File: package.json (hunk 1 above)
Comment (laljikanjareeya) [reply package.json]: Done !

File: package.json (hunk 2 above)
Comment (laljikanjareeya) [reply package.json]: Actually one was protos and other one was proto, but I don't think we need proto so removed.

File: src/transaction.ts (hunk 3 above)
Comment (laljikanjareeya) [reply src/transaction.ts]: Removed `RowCountsServiceError`

File: src/common.ts (hunk 5 above)
Comment (laljikanjareeya) [reply src/common.ts]: Done !

File: src/instance.ts (hunk 8 above)
Comment (laljikanjareeya) [reply src/instance.ts:99]: Done !

File: src/instance.ts (hunk 10 above)
Comment (laljikanjareeya) [reply src/instance.ts]: Done !

File: src/common.ts (hunk 11 above)
Comment (laljikanjareeya) [reply src/common.ts]: Removed `RowCountsServiceError` as it's not needed any more

File: src/common.ts (hunk 4 above)
Comment (laljikanjareeya) [reply src/common.ts]: Removed `RowCountsServiceError` as it's not needed any more

File: src/instance.ts (hunk 6 above)
Comment (laljikanjareeya) [reply src/instance.ts]: Done !

File: src/instance.ts (hunk 7 above)
Comment (laljikanjareeya) [reply src/instance.ts]: Done !

File: test/instance.ts [hunk 12]
```diff
@@ -61,13 +64,14 @@ describe('Instance', () => {
...
+  const sandbox = sinon.createSandbox();
 
-  const SPANNER = {
+  const SPANNER = ({
     request: util.noop,
     requestStream: util.noop,
     projectId: 'project-id',
     instances_: new Map(),
-  };
+  } as {}) as Spanner;
```
Comment (AVaksman) [inline test/instance.ts:74]: Do you need the round brackets `()` here?

**Review by AVaksman — APPROVED** [review]

File: test/instance.ts (hunk 12 above)
Comment (laljikanjareeya) [reply test/instance.ts:74]: it's added while running `npm run fix`

Comment (AVaksman) [reply test/instance.ts:74]: 👍

**Pull request closed without merging**